web/src/lib/data-hooks/useUnsavedChangesGuard.ts

```diff
@@ -0,0 +1,113 @@
+import { useRouter } from "next/compat/router";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+interface UseUnsavedChangesGuardProps {
+  hasUnsavedChanges: boolean;
+}
+
+interface UseUnsavedChangesGuardReturn {
+  isBlocked: boolean;
+  pendingUrl: string | null;
+  allowNavigation: () => void;
+  blockNavigation: () => void;
+  reset: () => void;
+}
+
+export const useUnsavedChangesGuard = ({
+  hasUnsavedChanges,
+}: UseUnsavedChangesGuardProps): UseUnsavedChangesGuardReturn => {
+  const router = useRouter();
+  const [isBlocked, setIsBlocked] = useState(false);
+  const [pendingUrl, setPendingUrl] = useState<string | null>(null);
+  const isNavigationAllowedRef = useRef(false);
+
+  const hasUnsavedChangesRef = useRef(hasUnsavedChanges);
+  hasUnsavedChangesRef.current = hasUnsavedChanges;
+
+  const allowNavigation = useCallback(() => {
+    isNavigationAllowedRef.current = true;
+    setIsBlocked(false);
+    if (pendingUrl && router) {
+      router.push(pendingUrl);
+    }
+    setPendingUrl(null);
+  }, [pendingUrl, router]);
+
+  const blockNavigation = useCallback(() => {
+    setIsBlocked(false);
+    setPendingUrl(null);
+    isNavigationAllowedRef.current = false;
+  }, []);
+
+  const reset = useCallback(() => {
+    setIsBlocked(false);
+    setPendingUrl(null);
+    isNavigationAllowedRef.current = false;
+  }, []);
+
+  useEffect(() => {
+    if (!router?.events || !router.isReady) return;
+
+    const handleRouteChangeError = (err: unknown): void => {
+      if ((err as { cancelled?: boolean })?.cancelled) {
+        return;
+      }
+      throw err as Error;
+    };
+
+    const handleRouteChangeStart = (url: string): void => {
+      if (!hasUnsavedChangesRef.current || isNavigationAllowedRef.current) {
+        isNavigationAllowedRef.current = false;
+        return;
+      }
+
+      setPendingUrl(url);
+      setIsBlocked(true);
+
+      if (process.env.NODE_ENV === "development") return;
+
+      const err: Error & { cancelled?: boolean } = new Error(
+        "Navigation blocked due to unsaved changes",
+      );
```

Comment:
I might actually add a comment explaining this piece. It's not a pattern I'm familiar with. Especially depending on the outcomes of our conversations tomorrow at engineering sync.